piecrust-uplink/src/lib.rs

```diff
@@ -4,13 +4,69 @@
 //
 // Copyright (c) DUSK NETWORK. All rights reserved.
 
+//! Piecrust Uplink is the crate for writing WASM smart-contracts in Rust
+//! targeting the Piecrust virtual machine.
+//!
+//! A smart-contract is a program exposing a collection of functions that
+//! operate on a shared state. These functions can perform a variety of
+//! operations, such as reading or modifying the state, or interacting with the
+//! host.
+//!
+//! # State Model and ABI
+//! Contracts targeting the Piecrust VM represent their state as a WASM memory.
+//! The contract is free to represent its data as it sees fit, and may allocate
+//! at will.
+//!
+//! To communicate with the host, both the contract and the host can
+//! emplace data in a special region of this memory called the argument buffer.
+//! The argument buffer is used as both the input from the host and as the
+//! output of the contract. All functions exposed by the contract must follow
+//! the convention:
+//!
+//! ```c
+//! // A function compatible with the piecrust ABI takes in the number of bytes
+//! // written by the host to the argument buffer and returns the number of
+//! // bytes written by the contract.
+//! uint32_t fname(uint32_t arg_len);
+//! ```
+//!
+//! The contract also has some functions available to it, offered by the host
+//! through WASM imports. Examples of these functions include, but are not
+//! limited to:
+//!
+//! - [`call`] to call another contract
+//! - [`emit`] to emit events
+//!
+//! The functions in this crate are wrappers around a particular way of calling
+//! the WASM imports. Take a look at the [externs] for a full view of what is
+//! available.
+//!
+//! # Examples
+//! Some of contracts used for testing purposes function as good example. Take a
```

Comment:
I would recommend changing this sentence to:
"Some of the contracts used for testing purposes function as good examples."